### DISCUSSION TITLE
docs/core/gdb-chaining-relation-with

### DISCUSSION THREAD

**Discussion by giscus** [gogf/gf-site#53] [General]
# docs/core/gdb-chaining-relation-with

GoFrame框架中的With特性，通过示例展示了如何使用With特性实现模型关联、数据查询，介绍了数据结构定义、事务操作、数据写入与查询等方面的内容，帮助开发者更好地理解和使用GoFrame框架进行高效开发。

https://goframe.org/docs/core/gdb-chaining-relation-with

**Comment by jeyrce** [comment]
相当于是作为 Join 的替代查询了，赞，建议在增加一下是否递归查询 With 的开关或者限制递归深度，毕竟如果子字段指定很多，将会发起很多次查询

**Reply by swift-fs** [reply]
不是Join的替代，SQL语句也不是Join查询，数据合并在代码层完成的，只能说是ScanList的简化，但是用下来还是喜欢ScanList一点

**Comment by wangle201210** [comment]
建议将案例中的gmeta.Meta改为g.Meta保持全局一致，否则看到有些地方是g.Meta有些地方是gmeta.Meta

**Comment by wangle201210** [comment]
dump出来的UserScores不应该是nil，应该是空数组‘[]’

**Comment by jsyzchen** [comment]
这种方式不适合不固定的order排序吧，比如UserDetail有age、city等字段，后端需要通过前端传的排序方式是age还是city来排序，需要获取User和UserDetail的列表

**Reply by gqcn** [reply]
是的，所以属于"静态关联"。

**Comment by hechen77** [comment]
```go
type UserScores struct {
    Id    int `json:"id"`
    Uid   int `json:"uid"`
    Score int `json:"score"`
}

type User struct {
    Id          int           `json:"id"`
    Name        string        `json:"name"`
}
```
User和UserScores属于`1:n`的关系，我想在查询User的时候顺带查出每个User对应的UserScores的数量，这个该怎么关联？

**Comment by miaoningzheng** [comment]
```go
type User struct {
    g.Meta `orm:"table:user"`
    Id         int           `json:"id"`
    Name       string        `json:"name"`
    UserDetail *UserDetail   `orm:"with:uid=id"`
    UserScores []*UserScores `orm:"with:uid=id"`
}
```
在`1:n`的场景下，我想限制`n`数量，比如上述中的 `UserScores` 我只想查出前`10`条，用`with`静态关联可以实现吗？

**Reply by gqcn** [reply]
> ```go
> type User struct {
>     g.Meta `orm:"table:user"`
>     Id         int           `json:"id"`
>     Name       string        `json:"name"`
>     UserDetail *UserDetail   `orm:"with:uid=id"`
>     UserScores []*UserScores `orm:"with:uid=id"`
> }
> ```
> 
> 在`1:n`的场景下，我想限制`n`数量，比如上述中的 `UserScores` 我只想查出前`10`条，用`with`静态关联可以实现吗？

你好，不行的。

**Reply by lankii** [reply]
同一张表里分组怎么实现递归关联一直失败
type DeviceGroupsNode struct {
	g.Meta      `orm:"table:device_groups, do:true"`
	Id          int         `json:"id"          orm:"id"          description:""`
	ParentId    string        `json:"parentId"    orm:"parent_id"   description:"默认0是父分组"`
	Name        string        `json:"name"        orm:"name"        description:"分组名称"`
	Description string        `json:"description" orm:"description" description:"描述"`
	Remark      string        `json:"remark"      orm:"remark"      description:""`
	TenantId    string        `json:"tenantId"    orm:"tenant_id"   description:"租户id"`
	
}



type DeviceGroupsList struct {
	g.Meta      `orm:"table:device_groups, do:true"`
	Id          int         `json:"id"          orm:"id"          description:""`
	ParentId    string        `json:"parentId"    orm:"parent_id"   description:"默认0是父分组"`
	Name        string        `json:"name"        orm:"name"        description:"分组名称"`
	Description string        `json:"description" orm:"description" description:"描述"`
	Remark      string        `json:"remark"      orm:"remark"      description:""`
	TenantId    string        `json:"tenantId"    orm:"tenant_id"   description:"租户id"`
	ChildNode   []DeviceGroupsNode `json:"child"  orm:"with:parent_id=id" description:"子分组"`
}

 g.Model("device_groups").With(entity.DeviceGroupsNode {}).Scan(&res)
打印出来child一直为空

**Reply by lankii** [reply]
err = g.Model("device_groups").With(entity.DeviceGroupsListNode{}.ChildNode).Scan(&res)  一样为空，看生成的sql只执行了一次select,
 [default]  [rows:4  ] SELECT * FROM `device_groups`
2025-03-19 11:40:46.744 {58f67db43a172e1898a9eb5f61c061b6} 200 "GET http 127.0.0.1:8808 /api/v1/device/group/list?parentId=03434

**Comment by notmaster-C** [comment]
这部分如何去避免硬编码

**Reply by notmaster-C** [reply]
这些字段和表名都需要自己去写，维护一个硬编码的文件成本不如维护一个sql文件吧

**Reply by lankii** [reply]
请问下，三个表关联查询应该怎么写，之前sql吗，如a,b,c三个表，  a:id,cfgid, b:id,tplId c:id,tplId   ,要根据a表中的a.cfgid=b.id  得到b表中的tplid,在关联c表中的b.tplid=c.tplId 中的数据，应该怎么写，纯sql. innerjion就好了

**Comment by popfaker** [comment]
感谢分享，请问在gf gen dao的情况下，系统会自动生成entity模型文件，默认是不需要编辑的，并且多次执行会自动覆盖。 用with方法这种硬编码关联方式，是否意味着就需要手动修改entity模型文件了？以后调整表结构都只能手动更新entity和do文件了？

**Reply by jeyrce** [reply]
永远不要修改自动生成的结构，如果需要进行这种关联，优先使用匿名嵌套entity的方法

**Reply by bravelib** [reply]
with说到底就是减少了代码量，完全可以不用with的

**Comment by zerowjz** [comment]
如果 where 条件需要多个参数要怎么写，比如  select * from table where (aa,bb)in(aaa,bbb),(ccc,ddd) 这样

**Comment by jsyzchen** [comment]
当前递归关联时会很慢，希望能够添加缓存，比如UserScoresOptional下面还有一个Course课程信息，在查询用户列表时，查询每个学生的分数，分数下又有Course课程信息，但是其实查询一个学生的分数时，所有课程信息都应该有了，查询第二个同学分数的时候可以直接使用内存缓存结果（虽然MySQL也有缓存，但比较慢），这样查询列表时会更快些，望采纳我的建议
